tests/cases/unittests/session.ts

```diff
@@ -0,0 +1,454 @@
+                const resultMsg = `Content-Length: ${len}\r\n\r\n${strmsg}\n`;
+
+                session.send = Session.prototype.send;
+                assert(session.send);
+                expect(session.send(msg)).to.not.exist;
+                expect(lastWrittenToHost).to.equal(resultMsg);
+            });
+        });
+
+        describe("addProtocolHandler", () => {
+            it("can add protocol handlers", () => {
+                const respBody = {
+                    item: false
+                };
+                const command = "newhandle";
+                const result = {
+                    response: respBody,
+                    responseRequired: true
+                };
+
+                session.addProtocolHandler(command, (req) => result);
+
+                expect(session.executeCommand({
+                    command,
+                    seq: 0,
+                    type: "command"
+                })).to.deep.equal(result);
+            });
+            it("throws when a duplicate handler is passed", () => {
+                const respBody = {
+                    item: false
+                };
+                const resp = {
+                    response: respBody,
+                    responseRequired: true
+                };
+                const command = "newhandle";
+
+                session.addProtocolHandler(command, (req) => resp);
+
+                expect(() => session.addProtocolHandler(command, (req) => resp))
+                .to.throw(`Protocol handler already exists for command "${command}"`);
+            });
+        });
+        
+        describe("event", () => {
+            it("can format event responses and send them", () => {
+                const evt = "notify-test";
+                const info = {
+                    test: true
+                };
+
+                session.event(info, evt);
+
+                expect(lastSent).to.deep.equal({
+                    type: "event",
+                    seq: 0,
+                    event: evt,
+                    body: info
+                });
+            });
+        });
+
+        describe("output", () => {
+            it("can format command responses and send them", () => {
+                const body = {
+                    block: {
+                        key: "value"
+                    }
+                };
+                const command = "test";
+
+                session.output(body, command);
+
+                expect(lastSent).to.deep.equal({
+                    seq: 0,
+                    request_seq: 0,
+                    type: "response",
+                    command,
+                    body: body,
+                    success: true
+                });
+            });
+        });
+    });
+
+    describe("how Session is extendable via subclassing", () => {
+        class TestSession extends Session {
+            lastSent: protocol.Message;
+            customHandler: string = "testhandler";

```

Comment:
Can we craft a tslint rule which checks for unnecessary types on inferrable literals? Seems like something we can do.